docs/manual/src/internals/async.md

```diff
@@ -0,0 +1,159 @@
+      - lift the return value
+      - send the return value into the `oneshot::Sender`.
+  - Call the callback method, passing a pointer to the function from the last step and the `oneshot::Sender` raw pointer.
+  - Return the receiver half of the `oneshot::Channel`.
+
+On the foreign side, the generated code simply runs the async function as normal, and call the callback once it completes.
+
+## Can you really just piggyback on the foreign async runtime?
+
+It's somewhat surprising, but yes.
+All the async code can run in the foreign runtime and Rust never has to start its own eventloop thread.
+In order to implement this, the Rust async functions must only await functions that fall into one of these categories:
+
+* UniFFI async callback/trait methods
+  This often means defining traits for things like HTTP clients so that Rust can leverage async HTTP libraries from the foreign side.
+* Async functions from crates like [oneshot](https://docs.rs/oneshot/latest/oneshot/) and [async_mutex](https://docs.rs/async-mutex/latest/async_mutex/) that work with any runtime.
+
+Beware that some Rust crates will present async APIs that silently start up runtimes in the background.
+For example, [reqwest](https://docs.rs/reqwest/latest/reqwest/) will start a tokio runtime.
+
+As an example of how it could work in the real-world, picture an API client.
+The Rust library depends an async HTTP callback interface to do the low-level fetching, then deserializes/validates the HTTP response.
+After that, it returns a high-level API response struct to the application.
+Furthermore, let's stipulate that the deserialization/validation is slow enough to be considered blocking.
+To manage this, the Rust library also depends on a (synchronous) callback interface that runs a task in a worker queue where blocking is allowed.
+Here's how the async API call could be handled:
+
+```mermaid
+sequenceDiagram
+  participant AsyncRust as Async Rust code
+  participant SyncRust as Sync Rust code
+  participant WorkerQueue as Work Queue Callback Interface
+  participant Http as Http Callback Interface
+  participant Foreign as Foreign language code
+
+  Foreign->>+AsyncRust: Start async API call
+  AsyncRust->>+Http: await method of Http callback interface
+  Http-->>-AsyncRust: Return HTTP response
+  Note over AsyncRust: Create oneshot channel<br/>and deseralization task
+  AsyncRust->>WorkerQueue: Schedule task
+  Note over AsyncRust: await oneshot channel
+  WorkerQueue->>SyncRust: Run task (in background thread)
+  SyncRust->>AsyncRust: Send result via oneshot channel
+  AsyncRust-->>-Foreign: Return API result
+```
+
+* The async Rust code runs in a [Future::poll](https://doc.rust-lang.org/std/future/trait.Future.html#tymethod.poll) method, inside the foreign event loop.  This code never blocks.
+* The sync Rust code runs inside a worker queue where it's okay to block.
+* This all combines together to create an async Rust call that's driven by the foreign event loop.
```

Comment:
Don't know where to put it: But I would love to also talk here about "data". Where is it? Showing the functions and who is calling what is excellent. **And** I would love to see some form of diagram which explains where does the data live and when does it getting transferred through the FFI etc. If that makes sense?

Reply:
I updated the diagram to do a bit more of this.  Initially, I had a very detailed diagram that showed all the FFI calls, but it got extremely busy so I decided to show more of a zoomed out view.